packages/concerto-core/lib/serializer/jsonpopulator.js

```diff
@@ -170,13 +170,64 @@ class JSONPopulator {
      * @private
      */
     visitMapDeclaration(mapDeclaration, parameters) {
-        const jsonObj = parameters.jsonStack.pop();
+        let jsonObj = parameters.jsonStack.pop();
         parameters.path ?? (parameters.path = new TypedStack('$'));
 
-        // throws if Map contains private reserved properties as keys.
+        // Throws if Map contains reserved properties - a Map containing reserved Properties should not be serialized.
         getAssignableProperties(jsonObj, mapDeclaration);
 
-        return new Map(Object.entries(jsonObj));
+        jsonObj = new Map(Object.entries(jsonObj));
+
+        let map = new Map();
+
+        jsonObj.forEach((value, key) => {
+
+            if (key === '$class') {
+                map.set(key, value);
+                return;
+            }
+
+            // If its a Non-Primitive, its likely a ClassDeclaration which needs visiting.
+            if (!ModelUtil.isSystemProperty(key) && !ModelUtil.isPrimitiveType(mapDeclaration.getKey().getType())) {
```

Comment:
Why do we check for system properties here? We already checked on line 177

Reply:
You're right, fixed.